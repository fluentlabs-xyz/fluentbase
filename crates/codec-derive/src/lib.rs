use proc_macro::TokenStream;
use syn::{self, Data, Fields};

fn impl_derive_codec(ast: &syn::DeriveInput) -> TokenStream {
    let data_struct = match &ast.data {
        Data::Struct(data_struct) => data_struct,
        _ => panic!("not supported yet"),
    };
    let named_fields = match &data_struct.fields {
        Fields::Named(named_fields) => named_fields,
        _ => panic!("not supported yet"),
    };
    for field in named_fields.named.iter() {
        // field.ty;
    }
    panic!("not supported yet")
}

#[proc_macro_derive(Codec)]
pub fn codec_macro_derive(input: TokenStream) -> TokenStream {
    let ast = syn::parse(input).unwrap();
    impl_derive_codec(&ast)
}

#[cfg(test)]
mod tests {
    use crate::impl_derive_codec;
    use quote::quote;

    #[test]
    #[ignore]
    fn test_simple_struct() {
        let input = quote! {
            struct Test {
                a: u32,
                b: u32,
                c: u32,
            }
        };
        let ast = syn::parse(input.into()).unwrap();
        impl_derive_codec(&ast);
    }
}
