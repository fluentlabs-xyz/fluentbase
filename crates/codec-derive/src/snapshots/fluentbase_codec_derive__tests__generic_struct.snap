---
source: crates/codec-derive/src/lib.rs
expression: get_generated_code(input)
---
impl<
    T,
    B: ::fluentbase_sdk::codec::byteorder::ByteOrder,
    const ALIGN: usize,
> ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { true }>
for GenericStruct<T, B, ALIGN>
where
    T: Clone + Default,
    T: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { true }>,
    Vec<T>: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { true }>,
{
    const HEADER_SIZE: usize = 0
        + <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::HEADER_SIZE
        + <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::HEADER_SIZE;
    const IS_DYNAMIC: bool = false
        || <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::IS_DYNAMIC
        || <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { true }>>::IS_DYNAMIC;
    fn encode(
        &self,
        buf: &mut ::fluentbase_sdk::codec::bytes::BytesMut,
        offset: usize,
    ) -> Result<(), ::fluentbase_sdk::codec::CodecError> {
        let aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let is_dynamic = <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::IS_DYNAMIC;
        let aligned_header_size = 0
            + if <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <T as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { true },
                    >>::HEADER_SIZE,
                )
            }
            + if <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <Vec<
                        T,
                    > as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { true },
                    >>::HEADER_SIZE,
                )
            };
        let mut tail = if is_dynamic {
            let buf_len = buf.len();
            let offset = if buf_len != 0 { buf_len } else { 32 };
            ::fluentbase_sdk::codec::write_u32_aligned::<
                B,
                ALIGN,
            >(buf, aligned_offset, offset as u32);
            if buf.len() < aligned_header_size + offset {
                buf.resize(aligned_header_size + offset, 0);
            }
            buf.split_off(offset)
        } else {
            if buf.len() < aligned_offset + aligned_header_size {
                buf.resize(aligned_offset + aligned_header_size, 0);
            }
            buf.split_off(aligned_offset)
        };
        let mut tail_offset = 0;
        if <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::IS_DYNAMIC {
            <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::encode(&self.field1, &mut tail, tail_offset)?;
            tail_offset += ::fluentbase_sdk::codec::align_up::<ALIGN>(4);
        } else {
            <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::encode(&self.field1, &mut tail, tail_offset)?;
            tail_offset
                += ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <T as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { true },
                    >>::HEADER_SIZE,
                );
        }
        if <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::IS_DYNAMIC {
            <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::encode(&self.field2, &mut tail, tail_offset)?;
            tail_offset += ::fluentbase_sdk::codec::align_up::<ALIGN>(4);
        } else {
            <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::encode(&self.field2, &mut tail, tail_offset)?;
            tail_offset
                += ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <Vec<
                        T,
                    > as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { true },
                    >>::HEADER_SIZE,
                );
        }
        buf.unsplit(tail);
        Ok(())
    }
    fn decode(
        buf: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<Self, ::fluentbase_sdk::codec::CodecError> {
        let mut aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let mut tmp = if <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::IS_DYNAMIC {
            let offset = ::fluentbase_sdk::codec::read_u32_aligned::<
                B,
                ALIGN,
            >(&buf.chunk(), aligned_offset)? as usize;
            &buf.chunk()[offset..]
        } else {
            &buf.chunk()[aligned_offset..]
        };
        let mut current_offset = 0;
        let field1 = <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::decode(&mut tmp, current_offset)?;
        current_offset
            += if <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <T as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { true },
                    >>::HEADER_SIZE,
                )
            };
        let field2 = <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::decode(&mut tmp, current_offset)?;
        current_offset
            += if <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { true },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <Vec<
                        T,
                    > as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { true },
                    >>::HEADER_SIZE,
                )
            };
        Ok(GenericStruct { field1, field2 })
    }
    fn partial_decode(
        buffer: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<(usize, usize), ::fluentbase_sdk::codec::CodecError> {
        let aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        if <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { true },
        >>::IS_DYNAMIC {
            let data_offset = ::fluentbase_sdk::codec::read_u32_aligned::<
                B,
                ALIGN,
            >(&buffer.chunk(), aligned_offset)? as usize;
            Ok((
                data_offset,
                <Self as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { true },
                    { true },
                >>::HEADER_SIZE,
            ))
        } else {
            Ok((
                aligned_offset,
                <Self as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { true },
                    { true },
                >>::HEADER_SIZE,
            ))
        }
    }
}
impl<
    T,
    B: ::fluentbase_sdk::codec::byteorder::ByteOrder,
    const ALIGN: usize,
> ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { false }>
for GenericStruct<T, B, ALIGN>
where
    T: Clone + Default,
    T: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { false }>,
    Vec<T>: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { true }, { false }>,
{
    const HEADER_SIZE: usize = 0
        + <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::HEADER_SIZE
        + <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::HEADER_SIZE;
    const IS_DYNAMIC: bool = false
        || <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC
        || <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC;
    fn encode(
        &self,
        buf: &mut ::fluentbase_sdk::codec::bytes::BytesMut,
        offset: usize,
    ) -> Result<(), ::fluentbase_sdk::codec::CodecError> {
        let aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let is_dynamic = <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC;
        let aligned_header_size = 0
            + if <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <T as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { false },
                    >>::HEADER_SIZE,
                )
            }
            + if <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <Vec<
                        T,
                    > as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { false },
                    >>::HEADER_SIZE,
                )
            };
        let mut tail = if is_dynamic {
            let buf_len = buf.len();
            let offset = if buf_len != 0 { buf_len } else { 32 };
            ::fluentbase_sdk::codec::write_u32_aligned::<
                B,
                ALIGN,
            >(buf, aligned_offset, offset as u32);
            if buf.len() < aligned_header_size + offset {
                buf.resize(aligned_header_size + offset, 0);
            }
            buf.split_off(offset)
        } else {
            if buf.len() < aligned_offset + aligned_header_size {
                buf.resize(aligned_offset + aligned_header_size, 0);
            }
            buf.split_off(aligned_offset)
        };
        let mut tail_offset = 0;
        if <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC {
            <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::encode(&self.field1, &mut tail, tail_offset)?;
            tail_offset += ::fluentbase_sdk::codec::align_up::<ALIGN>(4);
        } else {
            <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::encode(&self.field1, &mut tail, tail_offset)?;
            tail_offset
                += ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <T as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { false },
                    >>::HEADER_SIZE,
                );
        }
        if <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC {
            <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::encode(&self.field2, &mut tail, tail_offset)?;
            tail_offset += ::fluentbase_sdk::codec::align_up::<ALIGN>(4);
        } else {
            <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::encode(&self.field2, &mut tail, tail_offset)?;
            tail_offset
                += ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <Vec<
                        T,
                    > as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { false },
                    >>::HEADER_SIZE,
                );
        }
        buf.unsplit(tail);
        Ok(())
    }
    fn decode(
        buf: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<Self, ::fluentbase_sdk::codec::CodecError> {
        let mut aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let mut tmp = if <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC {
            let offset = ::fluentbase_sdk::codec::read_u32_aligned::<
                B,
                ALIGN,
            >(&buf.chunk(), aligned_offset)? as usize;
            &buf.chunk()[offset..]
        } else {
            &buf.chunk()[aligned_offset..]
        };
        let mut current_offset = 0;
        let field1 = <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::decode(&mut tmp, current_offset)?;
        current_offset
            += if <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <T as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { false },
                    >>::HEADER_SIZE,
                )
            };
        let field2 = <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::decode(&mut tmp, current_offset)?;
        current_offset
            += if <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { true },
                { false },
            >>::IS_DYNAMIC {
                32
            } else {
                ::fluentbase_sdk::codec::align_up::<
                    ALIGN,
                >(
                    <Vec<
                        T,
                    > as ::fluentbase_sdk::codec::Encoder<
                        B,
                        ALIGN,
                        { true },
                        { false },
                    >>::HEADER_SIZE,
                )
            };
        Ok(GenericStruct { field1, field2 })
    }
    fn partial_decode(
        buffer: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<(usize, usize), ::fluentbase_sdk::codec::CodecError> {
        let aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        if <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { true },
            { false },
        >>::IS_DYNAMIC {
            let data_offset = ::fluentbase_sdk::codec::read_u32_aligned::<
                B,
                ALIGN,
            >(&buffer.chunk(), aligned_offset)? as usize;
            Ok((
                data_offset,
                <Self as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { true },
                    { false },
                >>::HEADER_SIZE,
            ))
        } else {
            Ok((
                aligned_offset,
                <Self as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { true },
                    { false },
                >>::HEADER_SIZE,
            ))
        }
    }
}
impl<
    T,
    B: ::fluentbase_sdk::codec::byteorder::ByteOrder,
    const ALIGN: usize,
> ::fluentbase_sdk::codec::Encoder<B, ALIGN, { false }, { true }>
for GenericStruct<T, B, ALIGN>
where
    T: Clone + Default,
    T: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { false }, { true }>,
    Vec<T>: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { false }, { true }>,
{
    const HEADER_SIZE: usize = 0
        + ::fluentbase_sdk::codec::align_up::<
            ALIGN,
        >(
            <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { false },
                { true },
            >>::HEADER_SIZE,
        )
        + ::fluentbase_sdk::codec::align_up::<
            ALIGN,
        >(
            <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { false },
                { true },
            >>::HEADER_SIZE,
        );
    const IS_DYNAMIC: bool = false
        || <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::IS_DYNAMIC
        || <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::IS_DYNAMIC;
    fn encode(
        &self,
        buf: &mut ::fluentbase_sdk::codec::bytes::BytesMut,
        offset: usize,
    ) -> Result<(), ::fluentbase_sdk::codec::CodecError> {
        let mut current_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let header_size = <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::HEADER_SIZE;
        if buf.len() < current_offset + header_size {
            buf.resize(current_offset + header_size, 0);
        }
        <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::encode(&self.field1, buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <T as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { true },
                >>::HEADER_SIZE,
            );
        <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::encode(&self.field2, buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <Vec<
                    T,
                > as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { true },
                >>::HEADER_SIZE,
            );
        Ok(())
    }
    fn decode(
        buf: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<Self, ::fluentbase_sdk::codec::CodecError> {
        let mut current_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let field1 = <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::decode(buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <T as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { true },
                >>::HEADER_SIZE,
            );
        let field2 = <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { true },
        >>::decode(buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <Vec<
                    T,
                > as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { true },
                >>::HEADER_SIZE,
            );
        Ok(GenericStruct { field1, field2 })
    }
    fn partial_decode(
        buffer: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<(usize, usize), ::fluentbase_sdk::codec::CodecError> {
        let aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        Ok((
            aligned_offset,
            <Self as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { false },
                { true },
            >>::HEADER_SIZE,
        ))
    }
}
impl<
    T,
    B: ::fluentbase_sdk::codec::byteorder::ByteOrder,
    const ALIGN: usize,
> ::fluentbase_sdk::codec::Encoder<B, ALIGN, { false }, { false }>
for GenericStruct<T, B, ALIGN>
where
    T: Clone + Default,
    T: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { false }, { false }>,
    Vec<T>: ::fluentbase_sdk::codec::Encoder<B, ALIGN, { false }, { false }>,
{
    const HEADER_SIZE: usize = 0
        + ::fluentbase_sdk::codec::align_up::<
            ALIGN,
        >(
            <T as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { false },
                { false },
            >>::HEADER_SIZE,
        )
        + ::fluentbase_sdk::codec::align_up::<
            ALIGN,
        >(
            <Vec<
                T,
            > as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { false },
                { false },
            >>::HEADER_SIZE,
        );
    const IS_DYNAMIC: bool = false
        || <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::IS_DYNAMIC
        || <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::IS_DYNAMIC;
    fn encode(
        &self,
        buf: &mut ::fluentbase_sdk::codec::bytes::BytesMut,
        offset: usize,
    ) -> Result<(), ::fluentbase_sdk::codec::CodecError> {
        let mut current_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let header_size = <Self as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::HEADER_SIZE;
        if buf.len() < current_offset + header_size {
            buf.resize(current_offset + header_size, 0);
        }
        <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::encode(&self.field1, buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <T as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { false },
                >>::HEADER_SIZE,
            );
        <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::encode(&self.field2, buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <Vec<
                    T,
                > as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { false },
                >>::HEADER_SIZE,
            );
        Ok(())
    }
    fn decode(
        buf: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<Self, ::fluentbase_sdk::codec::CodecError> {
        let mut current_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        let field1 = <T as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::decode(buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <T as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { false },
                >>::HEADER_SIZE,
            );
        let field2 = <Vec<
            T,
        > as ::fluentbase_sdk::codec::Encoder<
            B,
            ALIGN,
            { false },
            { false },
        >>::decode(buf, current_offset)?;
        current_offset
            += ::fluentbase_sdk::codec::align_up::<
                ALIGN,
            >(
                <Vec<
                    T,
                > as ::fluentbase_sdk::codec::Encoder<
                    B,
                    ALIGN,
                    { false },
                    { false },
                >>::HEADER_SIZE,
            );
        Ok(GenericStruct { field1, field2 })
    }
    fn partial_decode(
        buffer: &impl ::fluentbase_sdk::codec::bytes::Buf,
        offset: usize,
    ) -> Result<(usize, usize), ::fluentbase_sdk::codec::CodecError> {
        let aligned_offset = ::fluentbase_sdk::codec::align_up::<ALIGN>(offset);
        Ok((
            aligned_offset,
            <Self as ::fluentbase_sdk::codec::Encoder<
                B,
                ALIGN,
                { false },
                { false },
            >>::HEADER_SIZE,
        ))
    }
}
